File: src/main/java/net/Chidoziealways/everythingjapanese/entity/client/triceratops/TriceratopsRenderer.java
package net.Chidoziealways.everythingjapanese.entity.client.triceratops;

import com.google.common.collect.Maps;
import com.mojang.blaze3d.vertex.PoseStack;
import net.Chidoziealways.everythingjapanese.EverythingJapanese;
import net.Chidoziealways.everythingjapanese.entity.TriceratopsVariant;
import net.Chidoziealways.everythingjapanese.entity.client.ModModelLayers;
import net.Chidoziealways.everythingjapanese.entity.custom.TriceratopsEntity;
import net.minecraft.Util;
import net.minecraft.client.model.AdultAndBabyModelPair;
import net.minecraft.client.renderer.MultiBufferSource;
import net.minecraft.client.renderer.entity.EntityRendererProvider;
import net.minecraft.client.renderer.entity.MobRenderer;
import net.minecraft.resources.ResourceLocation;

import java.util.Map;

public class TriceratopsRenderer extends MobRenderer<TriceratopsEntity, TriceratopsRenderState, TriceratopsModel<TriceratopsEntity>> {
    private static final Map<TriceratopsVariant, ResourceLocation> LOCATION_BY_VARIANT =
            Util.make(Maps.newEnumMap(TriceratopsVariant.class), map -> {
                map.put(TriceratopsVariant.GRAY,
                        ResourceLocation.fromNamespaceAndPath(EverythingJapanese.MOD_ID, "textures/entity/triceratops/triceratops_gray.png"));
                map.put(TriceratopsVariant.GREEN,
                        ResourceLocation.fromNamespaceAndPath(EverythingJapanese.MOD_ID, "textures/entity/triceratops/triceratops_green.png"));
            });
    private final AdultAndBabyModelPair<TriceratopsModel<TriceratopsEntity>> model;


    public TriceratopsRenderer(EntityRendererProvider.Context pContext) {
        super(pContext, new TriceratopsModel<>(pContext.bakeLayer(ModModelLayers.TRICERATOPS)), 0.86f);
        this.model = bakeModel(pContext);
    }

    private static AdultAndBabyModelPair<TriceratopsModel<TriceratopsEntity>> bakeModel(EntityRendererProvider.Context context) {
        TriceratopsModel<TriceratopsEntity> adult =
                new TriceratopsModel<>(context.bakeLayer(ModModelLayers.TRICERATOPS));
        TriceratopsModel<TriceratopsEntity> baby =
                new TriceratopsModel<>(context.bakeLayer(ModModelLayers.TRICERATOPS_BABY));
        return new AdultAndBabyModelPair<>(adult, baby);
    }


    @Override
    public ResourceLocation getTextureLocation(TriceratopsRenderState state) {
        return LOCATION_BY_VARIANT.get(state.getEntity().getVariant());
    }

    @Override
    public void render(TriceratopsRenderState pEntity, PoseStack pPoseStack,
                       MultiBufferSource pBuffer, int pPackedLight) {
        if(pEntity.isBaby()){
            pPoseStack.scale(0.5f, 0.5f, 0.5f);
        }else {
            pPoseStack.scale(1f, 1f, 1f);
        }

        super.render(pEntity, pPoseStack, pBuffer, pPackedLight);
    }

    @Override
    public TriceratopsRenderState createRenderState() {
        return new TriceratopsRenderState();
    }

    @Override
    public void extractRenderState(TriceratopsEntity p_368665_, TriceratopsRenderState p_363057_, float p_364497_) {
        p_363057_.setEntity(p_368665_);
        super.extractRenderState(p_368665_, p_363057_, p_364497_);
    }
}

File: src/main/kotlin/net/Chidoziealways/everythingjapanese/entity/client/triceratops/TriceratopsRenderer.kt
package net.Chidoziealways.everythingjapanese.entity.client.triceratops

import com.mojang.blaze3d.vertex.PoseStack
import net.Chidoziealways.everythingjapanese.JAPANESE_MOD_ID
import net.Chidoziealways.everythingjapanese.entity.TriceratopsVariant
import net.Chidoziealways.everythingjapanese.entity.client.ModModelLayers
import net.Chidoziealways.everythingjapanese.entity.custom.TriceratopsEntity
import net.minecraft.client.model.AdultAndBabyModelPair
import net.minecraft.client.renderer.MultiBufferSource
import net.minecraft.client.renderer.SubmitNodeCollector
import net.minecraft.client.renderer.entity.EntityRendererProvider
import net.minecraft.client.renderer.entity.MobRenderer
import net.minecraft.client.renderer.state.level.CameraRenderState
import net.minecraft.resources.Identifier
import java.util.*

class TriceratopsRenderer(pContext: EntityRendererProvider.Context) :
    MobRenderer<TriceratopsEntity, TriceratopsRenderState, TriceratopsModel<TriceratopsEntity>>(
        pContext,
        TriceratopsModel<TriceratopsEntity>(pContext.bakeLayer(ModModelLayers.TRICERATOPS)),
        0.86f
    ) {
    private val model: AdultAndBabyModelPair<TriceratopsModel<TriceratopsEntity>>


    init {
        this.model = bakeModel(pContext)
    }

    override fun getTextureLocation(state: TriceratopsRenderState): Identifier {
        return LOCATION_BY_VARIANT[state.entity!!.variant]!!
    }

    override fun submit(
        pEntity: TriceratopsRenderState, pPoseStack: PoseStack,
        collector: SubmitNodeCollector, cameraRenderState: CameraRenderState
    ) {
        if (pEntity.isBaby) {
            pPoseStack.scale(0.5f, 0.5f, 0.5f)
        } else {
            pPoseStack.scale(1f, 1f, 1f)
        }

        super.submit(pEntity, pPoseStack, collector, cameraRenderState)
    }

    override fun createRenderState(): TriceratopsRenderState {
        return TriceratopsRenderState()
    }

    override fun extractRenderState(p_368665_: TriceratopsEntity, p_363057_: TriceratopsRenderState, p_364497_: Float) {
        super.extractRenderState(p_368665_, p_363057_, p_364497_)
        p_363057_.entity = p_368665_
        p_363057_.idleAnimationState.copyFrom(p_368665_.idleAnimationState)
    }

    companion object {
        private val LOCATION_BY_VARIANT: Map<TriceratopsVariant, Identifier> by lazy {
            EnumMap<TriceratopsVariant, Identifier>(TriceratopsVariant::class.java).apply {
                put(
                    TriceratopsVariant.GRAY,
                    Identifier.fromNamespaceAndPath(
                        JAPANESE_MOD_ID,
                        "textures/entity/triceratops/triceratops_gray.png"
                    )
                )
                put(
                    TriceratopsVariant.GREEN,
                    Identifier.fromNamespaceAndPath(
                        JAPANESE_MOD_ID,
                        "textures/entity/triceratops/triceratops_green.png"
                    )
                )
            }
        }

        private fun bakeModel(context: EntityRendererProvider.Context): AdultAndBabyModelPair<TriceratopsModel<TriceratopsEntity>> {
            val adult =
                TriceratopsModel<TriceratopsEntity>(context.bakeLayer(ModModelLayers.TRICERATOPS))
            val baby =
                TriceratopsModel<TriceratopsEntity>(context.bakeLayer(ModModelLayers.TRICERATOPS_BABY))
            return AdultAndBabyModelPair<TriceratopsModel<TriceratopsEntity>>(adult, baby)
        }
    }
}
